zebrad/src/components/mempool/downloads.rs

```diff
@@ -298,4 +301,67 @@ where
 
         DownloadAction::AddedToQueue
     }
+
+    /// Check if transaction should be downloaded and verified.
+    ///
+    /// If it is already in the mempool (or in its rejected list)
+    /// or in state, then it shouldn't be downloaded (and an error is returned).
+    async fn should_download(
+        state: &mut ZS,
+        mempool: &mut ZM,
+        txid: UnminedTxId,
+    ) -> Result<(), BoxError> {
+        // Check if the transaction is already in the mempool.
+        match mempool
+            .ready_and()
+            .await?
+            .call(mp::Request::TransactionsById(
+                [txid].iter().cloned().collect(),
+            ))
+            .await
+        {
+            Ok(mp::Response::Transactions(txs)) => {
+                if txs.is_empty() {
+                    Ok(())
+                } else {
+                    Err("already present in mempool".into())
+                }
+            }
+            Ok(_) => unreachable!("wrong response"),
+            Err(e) => Err(e),
+        }?;
+
+        // Check if the transaction is in the mempool rejected list.
+        match mempool
+            .oneshot(mp::Request::RejectedTransactionIds(
+                [txid].iter().cloned().collect(),
+            ))
+            .await
```

Comment:
A Q: currently the rejected set includes victim tx's, that have been evicted only because a new tx came in and we only have a finite size pool. Is that ok?

Reply:
I think so. In https://zips.z.cash/zip-0401, when a transaction is received it must be dropped if it's in the evicted list ("If it is in RecentlyEvicted, the transaction MUST be dropped."). Should I add a comment mentioning that?

Reply:
Yeah re-reading ZIP-401 this should be fine.

Note to self: ZIP-401 recommends evicting a tx from the mempool basically at random, not just oldest, when we're out of room, I need to make a ticket for that.